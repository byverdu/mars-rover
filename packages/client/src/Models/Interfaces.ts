import { EnumCardinalPoints, EnumRoverStatus } from './enums';

export interface IPlateau {
  uuid: string;
  name: string;
  size: IArea;
  rovers: IRover[];
}

export interface IRover {
  uuid: string;
  uuidPlateau: string;
  lastKnownPosition: IRoverPosition;
  newPosition: IRoverPosition;
  stepsToNextPosition: {
    steps: string[];
    source: string;
  };
  status: EnumRoverStatus;
  dateCreation?: Date;
  modifiedAt?: Date;
}

export interface IRoverPosition {
  axis: ICoords;
  position: EnumCardinalPoints;
  rawFormat: string;
}

export interface IArea {
  width: number;
  height: number;
}

export interface ICoords {
  x: number;
  y: number;
}

export interface IRoverPositionPayload {
  steps: string;
  position: string;
}

export interface IPlateauPayload {
  plateauSize: string;
  rovers: IRoverPositionPayload[]
}
