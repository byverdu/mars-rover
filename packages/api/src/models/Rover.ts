import * as mongoose from 'mongoose';
import { IRover } from '../types/Interfaces';

async function preSaveHook(next) {
  if (this._doc) {
    let doc = <IRover>this._doc;
    let now = new Date();
    if (!doc.dateCreation) {
      doc.dateCreation = now;
    } else {
      doc.modifiedAt = now;
    }
  }
  next();
  return this;
}

const RoverSchema: mongoose.Schema = new mongoose.Schema({
  uuid: { type: String, required: true, unique: true },
  uuidPlateau: { type: String, required: true, unique: true },
  lastKnownPosition: {
    rawFormat: { type: String },
    position: { type: String, required: true },
    axis: {
      x: { type: Number, required: true },
      y: { type: Number, required: true }
    }
  },
  newPosition: {
    rawFormat: { type: String },
    position: { type: String },
    axis: {
      x: { type: Number },
      y: { type: Number }
    }
  },
  stepsToNextPosition: {
    steps: { type: Array, required: true },
    source: { type: String, required: true }
  },
  status: { type: String, required: true },
  dateCreation: { type: Date, required: false },
  modifiedAt: { type: Date, required: false }
});

RoverSchema.pre('save', preSaveHook);

// Export the model and return your IRover interface
export default mongoose.model<IRover>('Rover', RoverSchema);
