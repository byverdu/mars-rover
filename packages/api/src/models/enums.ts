export enum EnumCardinalPoints {
  N = 'N',
  S = 'S',
  E = 'E',
  W = 'W'
}

export enum EnumRoverStatus {
  moving = 'moving',
  sleep = 'sleep',
  ko = 'KO'
}

export enum EnumApiRoutes {
  getHealthCheck = '/health-check'
}
