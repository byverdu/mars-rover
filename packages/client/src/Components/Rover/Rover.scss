.rover {
  width: 50px;
  height: 50px;
  // border: 3px solid goldenrod;
  background: url(https://cdn-jr.brainpop.com/image_library/240000/240204_256x256.png)
    no-repeat center center fixed;
  background-size: cover;
}
