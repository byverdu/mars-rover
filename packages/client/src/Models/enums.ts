export enum EnumCardinalPoints {
  N = 'N',
  S = 'S',
  E = 'E',
  W = 'W'
}

export enum EnumsSteps {
  M = 'M',
  L = 'L',
  R = 'R'
}

export enum EnumRoverStatus {
  moving = 'moving',
  sleep = 'sleep',
  ko = 'KO'
}

export enum EnumApiRoutes {
  getHealthCheck = '/health-check',
  getPlateau = '/api/plateau',
  getRoot = '/',
  postPlateau = '/api/plateau/launch-rovers'
}

export enum EnumProxyAddress {
  development = 'http://localhost:9000',
  production = 'http://mars-rover_api:9000'
}

export enum EnumRotation {
  N = 'rotate-0',
  E = 'rotate-90',
  S = 'rotate-180',
  W = 'rotate-270'
}
