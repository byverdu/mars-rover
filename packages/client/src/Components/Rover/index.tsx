import React from 'react';
import './Rover.scss';
import { EnumCardinalPoints, EnumRoverStatus } from '../../Models/enums';

const classnames = require('classnames');

export interface RoverProps {
  axis: string;
  direction: EnumCardinalPoints;
  status: EnumRoverStatus;
};

const Rover: React.FC<RoverProps> = ({axis, direction, status}) => (
  <div
    className={classnames('rover', status)}
    data-axis={axis}
    data-direction={direction}
  />
);

export default Rover
