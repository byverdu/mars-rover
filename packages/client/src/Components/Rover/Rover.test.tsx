import React from 'react';
import Rover, { RoverProps } from './index';
import enzyme, { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { EnumCardinalPoints, EnumRoverStatus } from '../../Models/enums';

const props: RoverProps = {
  axis: 'x0:y0',
  direction: EnumCardinalPoints.N,
  status: EnumRoverStatus.sleep
};

describe('<Rover />', () => {
  const wrapper: enzyme.CommonWrapper = mount(<Rover {...props} />);

  it('should create a snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should know it's position on the grid", () => {
    expect((wrapper.props() as RoverProps).axis).toEqual('x0:y0');
  });

  it('should set the axis dataset attribute on the html element', () => {
    expect(
      wrapper.getDOMNode().attributes.getNamedItem('data-axis').value
    ).toEqual('x0:y0');
  });

  it('should know where is facing to', () => {
    expect((wrapper.props() as RoverProps).direction).toEqual(
      EnumCardinalPoints.N
    );
  });

  it('should set the direction dataset attribute on the html element', () => {
    expect(
      wrapper.getDOMNode().attributes.getNamedItem('data-direction').value
    ).toEqual('N');
  });

  it("should know where it's status", () => {
    expect((wrapper.props() as RoverProps).status).toEqual(
      EnumRoverStatus.sleep
    );
  });

  it('should set the direction dataset attribute on the html element', () => {
    expect(wrapper.getDOMNode().attributes.getNamedItem('class').value).toEqual(
      'rover sleep rotate-0'
    );
  });
});
