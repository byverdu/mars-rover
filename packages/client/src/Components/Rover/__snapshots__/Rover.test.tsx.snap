// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Rover /> should create a snapshot 1`] = `
<Rover
  axis="x0:y0"
  direction="N"
  status="sleep"
>
  <div
    className="rover sleep"
    data-axis="x0:y0"
    data-direction="N"
  />
</Rover>
`;
